### PR TITLE
VZ-5340 Changes for allowing prometheus to scrape metrics from authproxy

### PR DESCRIPTION
**Pull request by anajosep**
Signed-off-by: anajosep <anand.francis.joseph.augustin@oracle.com>

# Description

The goal of this task is to make `verrazzano-authproxy` metrics available in Prometheus. Since `verrazzano-authproxy`,  runs nginx as a standalone server, it requires a Prometheus exporter plugin ([nginx-prometheus-exporter](https://github.com/nginxinc/nginx-prometheus-exporter)) to be able to expose a metrics endpoint which can then be scraped by prometheus.

Fixes VZ-5340

# Checklist 

As the author of this PR, I have:

- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [x] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: platform-operator/controllers/verrazzano/component/authproxy/testdata/affinityOverrideValues.yaml [hunk 1]
```diff
@@ -2,6 +2,9 @@
 # Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
 imageName: ghcr.io/verrazzano/nginx-ingress-controller
 imageVersion: 0.46.0-20210510134749-abc2d2088
+metricsImageName: "ghcr.io/verrazzano/nginx-prometheus-exporter"
+metricsImageVersion: "0.10.0"
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/authproxy/testdata/affinityOverrideValues.yaml:6]: How is this image being built and why doesn't it follow the normal image naming pattern (e.g. 0.46.0-20210510134749-abc2d2088)

File: platform-operator/controllers/verrazzano/testdata/test_bom.json [hunk 2]
```diff
@@ -247,6 +247,12 @@
               "tag": "0.46.0-20210510134749-abc2d2088",
               "helmFullImageKey": "api.imageName",
               "helmTagKey": "api.imageVersion"
+            },
+            {
+              "image": "nginx-prometheus-exporter",
+              "tag": "0.10.0",
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/testdata/test_bom.json:253]: Do we have PLS approvals in place for this?

File: platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# Copyright (c) 2022, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+
+apiVersion: monitoring.coreos.com/v1
+kind: PodMonitor
+metadata:
+  name: authproxy
+  namespace: {{ .Release.Namespace }}
+  labels:
+    release: prometheus-operator
```
Comment (kminder) [inline platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml:10]: Is this required to have the Prometheus operator find this PodMonitor CR?

File: platform-operator/controllers/verrazzano/component/authproxy/testdata/affinityOverrideValues.yaml (hunk 1 above)
Comment (anajosep) [reply platform-operator/controllers/verrazzano/component/authproxy/testdata/affinityOverrideValues.yaml:6]: This image is being built using pulley and follows the version pattern defined there.

File: platform-operator/verrazzano-bom.json [hunk 4]
```diff
@@ -206,6 +206,12 @@
               "tag": "1.1.1-20220413170248-b60724ed1",
               "helmFullImageKey": "api.imageName",
               "helmTagKey": "api.imageVersion"
+            },
+            {
+              "image": "nginx-prometheus-exporter",
+              "tag": "0.10.0-test",
```
Comment (kminder) [inline platform-operator/verrazzano-bom.json:212]: What is this image tag with a -test suffix?

File: platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml (hunk 3 above)
Comment (davisgomes) [reply platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml:10]: Yes it is.

File: platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+spec:
+  namespaceSelector:
+    matchNames:
+      - {{ .Release.Namespace }}
+  selector: {}
```
Comment (davisgomes) [inline platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml:15]: I don't think you need to include this empty selector

File: platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  podMetricsEndpoints:
+    - port: https
+      path: /metrics
+      scheme: https
+      tlsConfig:
+        ca:
+          configMap:
```
Comment (wmhopkins) [inline platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml:22]: Shouldn't this be a secret instead of a config map?

**Review by wmhopkins — COMMENTED** [review]
Overall, looks good, I did have a few questions.

File: platform-operator/verrazzano-bom.json (hunk 4 above)
Comment (anajosep) [reply platform-operator/verrazzano-bom.json:212]: This is just for testing the image. Once the final release build is available, I will update this image tag to `0.10.0`

File: platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            key: root-cert.pem
+            name: istio-ca-root-cert
+            optional: true
```
[the remark this answers is not in the document]
Comment (anajosep) [reply platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml:25]: The configMap or secret for some of the TLS properties were not found, so I had made them optional and I didn't want to fail the auth proxy installation just due to missing configMap/secret.

File: platform-operator/verrazzano-bom.json (hunk 4 above)
Comment (anajosep) [reply platform-operator/verrazzano-bom.json]: The test image has been replaced with the released image.

File: platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml (hunk 5 above)
Comment (anajosep) [reply platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml:15]: Empty selector, gives a validation error during creation of PodMonitor/ServiceMonitor, so I am keeping it

File: platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (anajosep) [reply platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml]: Converted `PodMonitor` to `ServiceMonitor` as the Certs are available only as a file under `/etc/istio-certs` and `PodMonitor` supports only ConfigMap or Secret based configuration. With `ServiceMonitor` certs are being loaded from files at `/etc/istio-certs`

File: platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml (hunk 6 above)
Comment (anajosep) [reply platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml]: Converted `PodMonitor` to `ServiceMonitor` as the Certs are available only as a file under `/etc/istio-certs` and `PodMonitor` supports only ConfigMap or Secret based configuration. With `ServiceMonitor` certs are being loaded from files at `/etc/istio-certs`

File: platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        cert:
+          configMap:
```
[the remark this answers is not in the document]
Comment (anajosep) [reply platform-operator/helm_config/charts/verrazzano-authproxy/templates/monitors.yaml]: Converted `PodMonitor` to `ServiceMonitor` as the Certs are available only as a file under `/etc/istio-certs` and `PodMonitor` supports only ConfigMap or Secret based configuration. With `ServiceMonitor` certs are being loaded from files at `/etc/istio-certs`

**Review by fredtibbitts — APPROVED** [review]

**Review by stevshap — APPROVED** [review]

**Pull request merged**